python/lsst/meas/base/forcedPhotCcd.py

```diff
@@ -182,10 +234,72 @@ class ForcedPhotCcdTask(ForcedPhotImageTask):
     """
 
     ConfigClass = ForcedPhotCcdConfig
-    RunnerClass = lsst.pipe.base.ButlerInitializedTaskRunner
+    RunnerClass = pipeBase.ButlerInitializedTaskRunner
     _DefaultName = "forcedPhotCcd"
     dataPrefix = ""
 
+    def adaptArgsAndRun(self, inputData, inputDataIds, outputDataIds, butler):
+        # Filtering the reference catalog is currently handled by
+        # Gen2 specific methods.  To get around this, port in code
+        # segments to do the filtering and transformation.
+
+        # Step 1: Determine bounds of the exposure photometry will
+        # be performed on.
+        expWcs = inputData['exposure'].getWcs()
+        expRegion = inputData['exposure'].getBBox(lsst.afw.image.PARENT)
+        expBBox = lsst.geom.Box2D(expRegion)
+        expBoxCorners = expBBox.getCorners()
+        expSkyCorners = [expWcs.pixelToSky(corner).getVector() for
+                         corner in expBoxCorners]
+        expPolygon = lsst.sphgeom.ConvexPolygon(expSkyCorners)
+
+        # Step 2: Filter out reference catalog sources that are
+        # not contained within the exposure boundaries.
+        sources = type(inputData['refCat'])(inputData['refCat'].table)
+        for record in inputData['refCat']:
+            if expPolygon.contains(record.getCoord().getVector()):
+                sources.append(record)
+        refCatIdDict = {ref.getId(): ref.getParent() for ref in sources}
+
+        # Step 3: Cull sources that do not have their parent
+        # source in the filtered catalog.  Save two copies of each
+        # source.
+        sources = type(inputData['refCat'])(inputData['refCat'].table)
+        refSources = type(inputData['refCat'])(inputData['refCat'].table)
+        for record in inputData['refCat']:
+            if expPolygon.contains(record.getCoord().getVector()):
+                recordId = record.getId()
+                topId = recordId
+                while (topId > 0):
+                    if topId in refCatIdDict:
+                        topId = refCatIdDict[topId]
+                    else:
+                        break
+                if topId == 0:
+                    sources.append(record)
+                    refSources.append(record)
+
+        # Step 4: Transform source footprints from the reference
+        # coordinates to the exposure coordinates.
+        for srcRecord, refRecord in zip(sources, refSources):
+            srcRecord.setFootprint(refRecord.getFootprint().transform(inputData['refWcs'],
+                                                                      expWcs, expRegion))
+        # Step 5: Replace reference catalog with filtered source list.
+        inputData['refCat'] = sources
+
+        if 'measCat' not in inputData.keys():
```

Comment:
Would you add a comment saying why you're adding `measCat` to the inputs when its an output?  When do we expect it to be in  `inputData.keys()`?